### ISSUE TITLE
srt_getpeername get wrong result.

### ISSUE THREAD

**Issue by Edward-Wu** [Haivision/srt#683]
hi, I want to get the peer name by srt_getpeername, but I get wrong name. So I search the system function getpeername in the srt source code, I find out which is called by  CChannel::getPeerAddr function. 
But the CChannel::getPeerAddr is not called by each function.

Is there any one encounter such a thing?

-------
void CChannel::getPeerAddr(sockaddr* addr) const
{
   socklen_t namelen = m_iSockAddrSize;
   ::getpeername(m_iSocket, addr, &namelen);
}

**Comment by ethouris** [comment]
You are searching from the wrong direction. See `api.cpp` file, `CUDTUnited::getpeername`.

The "peer name" should be the address of the opposite side of the connection. It's always returned the same way. How it's set, it depends on whether the socket is a socket you use with `srt_connect` or returned by `srt_accept`.

**Comment by Edward-Wu** [comment]
hi, ethouris, thanks for your advice, I want get the peer name after srt_accept function at the server side.
According by your note, I follow the CUDTUnited::getpeername function, then I get the following processes:

1.CUDTUnited::getpeername call the CUDTUnited::getpeername, the second parameter 'sockaddr * name' is our clue. 

2.In CUDTUnited::getpeername function, I find out:
memcpy(name, s->m_pPeerAddr, *namelen);
so I trace the member m_pPeerAddr，I find out which is assigned in CDUTUnited::newConnection function.
memcpy(ns->m_pPeerAddr, peer, sizeof(sockaddr_in));
so I trace the parameter peer .

3.the parameter peer is transfered through CDUTUnited::newConnection, 
int CUDTUnited::newConnection(const SRTSOCKET listen, const sockaddr* peer, CHandShake* hs, const CPacket& hspkt)

4. newConnection is called by processConnectRequest :
int CUDT::processConnectRequest(const sockaddr* addr, CPacket& packet)
{
......
       int result = s_UDTUnited.newConnection(m_SocketID, addr, &hs, packet);
......
}
parameter peer change to addr.

5.processConnectRequest is called by CRcvQueue::worker_ProcessConnectionRequest .
CRcvQueue::worker_ProcessConnectionRequest(CUnit* unit, const sockaddr* addr)
{
......
       listener_ret = m_pListener->processConnectRequest(addr, unit->m_Packet);
......
}

6.worker_ProcessConnectRequest is called by CRcvQueue::worker, which is a thread function.
void* CRcvQueue::worker(void* param)
{
......
           if (id == 0)
           {
               cst = self->worker_ProcessConnectionRequest(unit, &sa);
           }
           else
           {
               cst = self->worker_ProcessAddressedPacket(id, unit, &sa);
           }
......
}

here parameter addr change to sa.
and I notice the sa is from the previous called function worker_RetrieveUnit, so I continue to trace the worker_RetrieveUnit function.

7. CRcvQueue::worker_RetrieveUnit call the CChannel::recvfrom with parameter addr.
        EReadStatus rst = m_pChannel->recvfrom(addr, temp);

8.in CChannel::recvfrom function, addr is transfered to system function recvfrom as the member of msghdr.msg_name.

Now, I don't know as a server side how to get the the address of the opposite side of the connection if we don't call the system function ::getpeername , especially if the connector is behind in the NAT which don't know the internet IP of itself.

Can your give more advices?

**Comment by Edward-Wu** [comment]
hi, ethouris, thank you. 
I know the system function getpeeraddr is used in tcp , otherwise in udp the peer name is transfered by the recvfrom or recvmsg function, so you are right. I will continue to trace why don't get the right peer name in my code .
Thanks again.

**Comment by Edward-Wu** [comment]
hi, ethouris,
I add some log info to srt source code, and use the example srt-live-transmit as a server, then the result of calling srt_getpeername is right.
Because my code refers to srt of FFMPEG, so I add some log info in ffmpeg source code with --enable-libsrt, code likes following:

static int libsrt_setup(URLContext *h, const char *uri, int flags)
{
......
    struct sockaddr_in peer_addr;
    int peer_len =  sizeof(peer_addr);
    char peer_name[30] = {0};
......
        if ((ret = libsrt_listen(s->eid, fd, cur_ai->ai_addr, cur_ai->ai_addrlen, h, open_timeout / 1000)) < 0)
            goto fail1;
        fd = ret;

        //following code for peer name log
        ret = srt_getpeername(fd, (struct sockaddr *)&peer_addr, &peer_len);
        if (0 == ret) {
            av_log(h, AV_LOG_ERROR,"libsrt_listen, new client, %s: %d\n",
                inet_ntop(AF_INET, &peer_addr.sin_addr, peer_name, sizeof(peer_name)),
                ntohs(peer_addr.sin_port));
        }
......
}

then make ffmpeg, then start my test:
1. ffmpeg as server and srt-live-transmit as client:
$ ./ffmpeg -re -i ../../../movie/basketball/BeiShenDanDa.flv -vcodec copy -acodec copy  -pkt_size 1316 -flush_packets 0 -f mpegts "srt://:8080?mode=listener"

and at the same computer, run the srt-live-transmit:
$ ./srt-live-transmit srt://127.0.0.1:8080 udp://127.0.0.1:5000

unluckly , I get the same wrong peer name in log info:
libsrt_listen, new client, 0.0.0.0: 60763

2.ffmpeg as client and srt-live-tranmit as server
$./ffmpeg -re -i ../../../movie/basketball/BeiShenDanDa.flv -vcodec copy -acodec copy  -pkt_size 1316 -flush_packets 0 -f mpegts "srt://127.0.0.1:8080"

and at the same computer, run the srt-live-transmit:
$./srt-live-transmit srt://:8080?mode=listener udp://127.0.0.1:5000

this time, I get the right peer name in log info:
recvmsg from:0x70000f9b8e78:127.0.0.1:63575, size=1332

I don't know who is the code's author of srt in ffmpeg. Is it possible that there is a bug with srt code in ffmpeg?

**Comment by ethouris** [comment]
Ok, let me continue the analysis beyond the call to `worker_ProcessConnectionRequest`. This is called with `&sa`, which is first filled earlier, at the call to `worker_RetrieveUnit`. This is then passed as a variable to fill the `msghdr` structure that is then passed to the call of `::recvmsg`.

So, in other words, this is the address from which there have come the packet, which is then recognized as a handshake response type `URQ_CONCLUSION`, which triggers creation of the new connection through the call to `CUDTUnited::newConnection`. This address is passed unchanged from there. Finally in `newConnection` it should be rewritten to `m_pPeerAddr` field.

If you suspect any scenario in which the address wouldn't be correct, you can try to trace what happens to this parameter since the call to `recvmsg` up to `newConnection` and see if this value is anywhere displayed in the logs. If not, you can add some logs, like:

`HLOGC(mglog.Debug, log << "PEER ADDRESS: " << SockaddrToString(addr));`

Just to see if this was really set correctly in the beginning and wasn't changed later.

**Comment by Edward-Wu** [comment]
hi ethouris, thanks for your analysis.
I trace the recvmsg function, and find out the value of member m_iSockAddrSize is not 16 but 28. 
m_iSockAddrSize is inited in CChannel::CChannel function:

m_iSockAddrSize = (AF_INET == m_iIPversion) ? sizeof(sockaddr_in) : sizeof(sockaddr_in6);

so the m_iSockAddrSize is only relative with the version of IPvX.
Finnally, I found out the problem, which's in libsrt_setup function in FFMPEG code:

    hints.ai_family = AF_UNSPEC;

I changed it to following:
    hints.ai_family = AF_INET;//AF_UNSPEC;
All's right!

thanks again.
